### PR TITLE
added wofSSL_EVP_PKEY_param_check for compatibility layer API

### PR DESCRIPTION
**Pull request by miyazakh**
EVP_PKEY_param_check compatibility layer API

### CODE REVIEW

File: tests/api.c [hunk 1]
```diff
@@ -36699,6 +36699,60 @@ static void test_wolfSSL_EVP_PKEY_assign_DH(void)
...
+#if !defined(NO_DH) && !defined(NO_FILESYSTEM)
+#if !defined(HAVE_FIPS) || (defined(HAVE_FIPS_VERSION) \
+                                     && (HAVE_FIPS_VERSION>2))
+    DH       *dh    = NULL;
+    DH       *setDh = NULL;
+    EVP_PKEY *pkey  = NULL;
+    EVP_PKEY_CTX*   ctx = NULL;
+    
+    FILE* f = NULL;
+    unsigned char buf[4096];
```
Comment (cconlon) [inline tests/api.c:36714]: Does this array need to be 4096?  Looks like we read dh2048.der into it, but that file is only 268 bytes in size.

File: wolfcrypt/src/evp.c [hunk 2]
```diff
@@ -2102,6 +2102,148 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_cmp(const WOLFSSL_EVP_PKEY *a, const WOLFSSL_EV
...
+        WOLFSSL_MSG("failed to assign big number");
+        ret = WOLFSSL_FAILURE;
+    }
+    
+    /* prime check */
+    if (ret == WOLFSSL_SUCCESS &&
+        wolfSSL_BN_is_odd(dh_key->p) == 0){
+        WOLFSSL_MSG("dh_key->p is not prime");
+        ret = WOLFSSL_FAILURE;
+    } /* TO DO safe prime check. need BN_rshift1 */
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:2137]: "TO DO" should be "TODO" to match other occurrences in code and easier to find later.

File: wolfcrypt/src/evp.c [hunk 3]
```diff
@@ -2102,6 +2102,148 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_cmp(const WOLFSSL_EVP_PKEY *a, const WOLFSSL_EV
...
+                WOLFSSL_MSG("dh_key->g is not suitable generator");
+                ret = WOLFSSL_FAILURE;
+            }
+#ifdef WOLFSSL_KEY_GEN
+        /* test if the number q is prime. */
+        if (ret == WOLFSSL_SUCCESS &&
+            (wolfSSL_BN_is_prime_ex(dh_key->q, 64, ctx, NULL) <= 0)) {
+            WOLFSSL_MSG("dh_key->q is not prime or error during check.");
+            ret = WOLFSSL_FAILURE;
+        } /* else TO DO check q div q - 1. need BN_div */
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:2175]: Same as above, "TO DO" should be "TODO"

File: wolfcrypt/src/evp.c [hunk 4]
```diff
@@ -2102,6 +2102,148 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_cmp(const WOLFSSL_EVP_PKEY *a, const WOLFSSL_EV
...
+    
+    /* sanity check */
+    if (ctx == NULL) {
+        return WOLFSSL_FAILURE;
+    }
+    
+    type = wolfSSL_EVP_PKEY_type(wolfSSL_EVP_PKEY_base_id(ctx->pkey));
+    
+    switch(type) {
+        #if !defined(NO_RSA)
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:2208]: Back if statements 4 spaces to make easier to read, thanks

File: wolfcrypt/src/evp.c [hunk 5]
```diff
@@ -2102,6 +2102,148 @@ WOLFSSL_API int wolfSSL_EVP_PKEY_cmp(const WOLFSSL_EVP_PKEY *a, const WOLFSSL_EV
...
+        case EVP_PKEY_RSA:
+            WOLFSSL_MSG("not yet implemented");
```
Comment (cconlon) [inline wolfcrypt/src/evp.c:2210]: Let's make these messages slightly more specific by mentioning the type in them.  For example "EVP_PKEY_RSA not yet implemented".  Applies to all similar cases below.

**Review by cconlon — CHANGES_REQUESTED** [review]

File: tests/api.c (hunk 1 above)
Comment (miyazakh) [reply tests/api.c]: That's true. Changed to 512 bytes for buffer.

File: wolfcrypt/src/evp.c (hunk 2 above)
Comment (miyazakh) [reply wolfcrypt/src/evp.c]: Fixed

File: wolfcrypt/src/evp.c (hunk 3 above)
Comment (miyazakh) [reply wolfcrypt/src/evp.c]: Fixed

File: wolfcrypt/src/evp.c (hunk 4 above)
Comment (miyazakh) [reply wolfcrypt/src/evp.c:2219]: Added 4 spaces. Thanks.

File: wolfcrypt/src/evp.c (hunk 5 above)
Comment (miyazakh) [reply wolfcrypt/src/evp.c]: Thanks for the example. Added type to each cases.

**Review by cconlon — APPROVED** [review]

**Pull request merged**